### PR TITLE
Stabilise disable minification

### LINKED ISSUE #54819: [rustdoc] stabilise disable-minification option

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #54819.
Reopening of #54827.

r? @QuietMisdreavus

### CODE REVIEW

**Comment by ollie27** [comment]
What is the use case for this option? Surely if people want to see the none minified versions they can just look at rustdoc's source code.

**Comment by GuillaumeGomez** [comment]
Well, rustdoc messes with JS content by adding stuff in it. So it's not that simple to get a view of what's really generated unless you disable the minification.

**Comment by QuietMisdreavus** [comment]
Rustdoc also minifies any user-supplied themes as well as the search index. It also adds an extra line to `storage.js` with the resource suffix, before minifying that.

This is a debugging flag that @GuillaumeGomez uses personally, and the original PR for this was among a handful of no-comment stabilization PRs that he started at the same time, alongside threads for `--themes` and `--resource-suffix`.

It looks like you've added docs since the original PR, but i'm not aware of any tests in this repo for the minifier functionality. (How much is the `minifier` crate itself tested? That could offer a stand-in for testing the *functionality* of this feature, though not the flag.) I still think it would be good to have a `run-make-fulldeps` test that passes this flag to generate docs for a demo crate (probably an empty crate, but it could have anything), then makes sure that `rustdoc.css` and `main.js` have the same file contents as their original files in the repo.

Regardless, we should hold an FCP before merging this, since it's a feature stabilization.

**Comment by GuillaumeGomez** [comment]
Testing is complicated directly on rustdoc since everytime there is an update, we'll have to update the minification tests as well. I added tests (maybe not enough?) in the minifier crate directly. I'll let you judge if there are enough, and if not, I'll add even more! :D

**Comment by Centril** [comment]
Ping from triage, @QuietMisdreavus and @GuillaumeGomez, any progress?

**Comment by GuillaumeGomez** [comment]
Well, I'm still waiting for a definitive opinion. I'll rebase though.

**Comment by GuillaumeGomez** [comment]
Need to check on gtk-rs if cargo workspaces could fix my issue.

**Comment by QuietMisdreavus** [comment]
We discussed this at the Rust All-Hands. @GuillaumeGomez discussed a use case for this in gtk-rs, where there were difficulties in writing all the docs at once. However, it seems like putting them into a Cargo workspace could solve the issue, so imperio needs to investigate that. If that fixes that issue, the only remaining use case is debugging, and it's not pressing to have a debugging flag stabilized. If that's the only remaining use, i would move to close this PR.

**Comment by Dylan-DPC** [comment]
ping from triage @QuietMisdreavus waiting for your review on this

**Comment by QuietMisdreavus** [comment]
I still want to see whether @GuillaumeGomez was able to solve the issues in gtk-rs with workspaces instead of using this flag. If workspaces work for gtk-rs, then i don't think this needs to be stabilized.

**Comment by GuillaumeGomez** [comment]
I didn't take time to do it yet. ><

**Comment by TimNN** [comment]
Ping from triage @GuillaumeGomez: What is the status of this PR?

**Comment by GuillaumeGomez** [comment]
I still need to check on gtk-rs...

**Comment by Dylan-DPC** [comment]
ping from triage @GuillaumeGomez any updates? is this blocked on anything?

**Comment by GuillaumeGomez** [comment]
I still need to check the workspace story but it'll need to wait for a gtk-rs update which is taking longer than expected.

**Comment by Centril** [comment]
@GuillaumeGomez Any updates?

**Comment by GuillaumeGomez** [comment]
I can't do it on gtk-rs so the current PR is in its "final" state.

**Comment by Centril** [comment]
@QuietMisdreavus And given the PR in its final state ^--, how shall we proceed? (ping from triage... etc. etc.)

**Comment by GuillaumeGomez** [comment]
I honestly don't know. It didn't happen much times in the rustdoc team! :)

File: src/doc/rustdoc/src/unstable-features.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/unstable-features.md]:
```suggestion
By default, all JS and CSS files are minified in order to make them smaller. However,
sometimes there are bugs in the minification process, or the files are harder to read
when this happens. Passing this flag to rustdoc skips the minifier, passing the original
JS and CSS all the way into the final output.
```

File: src/doc/rustdoc/src/unstable-features.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/unstable-features.md]: This should be in the "Command-line arguments" chapter, not the "Unstable features" chapter, if we're stabilizing the flag.

**Review by QuietMisdreavus — CHANGES_REQUESTED** [review]
From what i can tell, there are a couple potential uses for this flag:

* Debugging search index/aliases/etc output while developing rustdoc itself.
* Working around bugs in the minifier.

If it were just the first one (as i thought when this PR was initially posted), then i would be opposed to this change (as i was back then). However, the performance of the minifier has slowed down a lot since it was initially introduced, due to the introduction of more features. In addition, certain project structures like in gtk-rs cause issues when `cargo doc` is run multiple times into the same output directory with different packages. Regardless, this flag has had some usage already, so the flag itself should be fine. (I'll try to experiment with adding a test for the flag, since i think that we should at least have some kind of verification that the flag does what we want it to.)

**Comment by QuietMisdreavus** [comment]
@rfcbot fcp merge

This is a stabilization of the `--disable-minification` CLI flag in rustdoc, which disables the minifier that is usually run on JavaScript and CSS files embedded with documentation builds. Rationale for stabilization is included in my review comment immediately above this one.

**Comment by GuillaumeGomez** [comment]
@QuietMisdreavus I applied the changes you suggested.

**Comment by QuietMisdreavus** [comment]
I was able to write a small test to ensure that the `--disable-minification` flag runs the minifier as expected:

```rust
// compile-flags:--disable-minification

// @has search-index.js
// @!has - 'var R'

/// The quick brown fox jumps over the lazy dog.
pub struct SomeStruct;

/// The quick brown fox jumps over the lazy dog.
pub fn some_fn() {}
```

This just checks for the `R` string array that the minifier adds to the search index. There need to be some items in the crate, otherwise the minifier doesn't have anything useful to run on and skips collapsing the strings. This has the risk of being flaky if the minifier changes later on, but we can at least be sure that it's running in the first place.

**Review by QuietMisdreavus — APPROVED** [review]

**Comment by QuietMisdreavus** [comment]
I'd like to see if @ollie27 wants to weigh in, but otherwise, if you add that test it LGTM.

**Comment by GuillaumeGomez** [comment]
I'll add your test.

**Comment by GuillaumeGomez** [comment]
Test added!

**Comment by QuietMisdreavus** [comment]
Thanks! Now i just want to wait for ollie27's comment, and we can ride the FCP out.

**Comment by ollie27** [comment]
What exactly are the issues with the minifier? I am aware of #58849 and I've just opened #61216 but the way I see it, if there are bugs with this feature then it should be fixed or removed. It's just an optimization after all.

**Comment by euclio** [comment]
The issues with the minifier stem from the fact that it is run on every invocation of rustdoc, instead of after all the documentation has been collected.

For example, since the JavaScript minification only adds to the minified string index (the `R` array), it's sensitive to the order that the crates are documented. This is fixable by modifying the minifier to re-read the contents `R` if present, add any new string candidates, sort them, and then fix up any existing `R` indexing operations to use the new indices. This is complex, and also requires that the minifier has special knowledge of the variables used for minification.

Even with this fixed, there's still a lot of useless work going on per-invocation. A simple solution to both issues would be to wait for all the rustdoc invocations to finish, and then run the minifier once over all the artifacts. However, I think this would require `cargo` integation, since there otherwise isn't a way to know that all the `rustdoc` invocations have finished.

However, if minification happened after `rustdoc` finishes, it would be possible to run a mature, off-the-shelf minifier instead of the built-in one. Third-party minifiers likely offer better minification with fewer bugs. But, it wouldn't be automatic or included for all users.

I understand that saving bandwidth is important, especially to users who have poor internet connections. But, I'm skeptical that minification is `rustdoc`'s responsibility. How much bandwidth/disk space is actually saved? Does it justify the performance cost to all users of `rustdoc`, including those running it locally? If minification *is* necessary, I think offloading minification to a step before deploying the documentation in CI or docs.rs is better than minifying inside `rustdoc`.

**Comment by GuillaumeGomez** [comment]
@euclio: Considering the size of the `search-index.js` file, I think it's necessary. Give it a try on a big library without the minification enabled and you'll see. Otherwise I agree about the string indexation: it should re-read the `R` index and remake the job.

**Comment by QuietMisdreavus** [comment]
I think over time i've come around to @euclio's position. I'm aware that rustdoc generates a large file for the search index, but that doesn't stop it from [being a performance bottleneck](https://github.com/rust-lang/rust/issues/58849) or occasionally generating [faulty](https://github.com/rust-lang/rust/pull/59958) or [inconsistent](https://github.com/rust-lang/rust/issues/61216) JS.

I believe our niche as a tool doesn't need to include making sure that we serve optimally-small resources. If some site hosting docs wishes to shrink the resources it loads, it can minify and/or gzip them after the fact. This means we waste less time in the average case (user generating local docs for themselves). I feel like the best place for the minifier is as an external tool that can be called as a separate script - `cargo minidoc` perhaps - that minifies resources separately. When built into rustdoc and called repeatedly, it seems to create more problems than it solves.

**Comment by Dylan-DPC** [comment]
Closing this after a discussion with @GuillaumeGomez

**Pull request closed without merging**
